feat: let plugins be functions with config access

csr632 opened the pull request:
Provide a plugin hook to access and alter vite config. It imitates rollup's [options](https://rollupjs.org/guide/en/#options) hook.

Usage:
```ts
import { Plugin } from 'vite'

export const changeConfigPlugin1: Plugin = {
  changeConfig: (prev) => {
    prev.define['__changeConfigPlugin1__'] = 'success'
    prev.plugins.push(changeConfigPlugin2)
    return prev
  }
}

const changeConfigPlugin2: Plugin = {
  changeConfig: (prev) => {
    prev.define['__changeConfigPlugin2__'] = 'success'
    return prev
  }
}
```

fix: https://github.com/vitejs/vite/issues/1162

csr632 committed:
feat: plugin changeConfig hook

aleclarson reviewed (changes requested):
Can we move this into `resolveConfig` so it gets cached by #740?

aleclarson commented:
And maybe `editConfig` has a nicer ring to it? 🚲

csr632 commented:
> I wonder if we should even let it return another config? It's faster to mutate than copy, but maybe no big deal?

This hook already allow you to return another config. Checkout this line: https://github.com/vitejs/vite/pull/1178/files#diff-e1f89aee8992fa7022cc43adb93042f9bbb4b223f0d6cd0e70719cf4b795afa4R164

csr632 commented:
> And maybe `prepareConfig` has a nicer ring to it?

What does "ring" mean? Does it mean "name"?  :)
This hook is imitating rollup's [options](https://rollupjs.org/guide/en/#options) hook. How about name it `options`?

csr632 commented:
> Can we move this into `resolveConfig` so it gets cached by #740?

I voted against that pr, checkout https://github.com/vitejs/vite/pull/740#issuecomment-736480556

aleclarson commented:
> > Can we move this into `resolveConfig` so it gets cached by #740?
> 
> I voted against that pr, checkout [#740 (comment)](https://github.com/vitejs/vite/pull/740#issuecomment-736480556)

Okay, but we should still move it into `resolveConfig`, I think.

> > And maybe `prepareConfig` has a nicer ring to it?
> 
> What does "ring" mean? Does it mean "name"? :)

https://www.gymglish.com/en/gymglish/english-translation/a-nice-ring-to-it

> This hook is imitating rollup's [options](https://rollupjs.org/guide/en/#options) hook. How about name it `options`?

I'm not a big fan of using "config" and "options" interchangeably. I think we should stick with "config" naming with a verb like "change" or "prepare" before it. Just naming it `config` might give the wrong impression.

> > I wonder if we should even let it return another config? It's faster to mutate than copy, but maybe no big deal?
> 
> This hook already allow you to return another config. Checkout this line: [#1178 (files)](https://github.com/vitejs/vite/pull/1178/files#diff-e1f89aee8992fa7022cc43adb93042f9bbb4b223f0d6cd0e70719cf4b795afa4R164)

Sorry, I worded that confusingly. I meant we should consider *not* allowing a new config to be returned. Two reasons to do this: (1) so other plugins can keep a stable reference to the Vite config, and (2) it's at least slightly more performant to mutate the config than to copy it.

aleclarson committed:
feat: let plugins be functions with config access

When a plugin is a function, it receives a stable reference to the Vite config. To make this work, we should avoid cloning the config.

aleclarson committed:
fix: merge argv instead of cloning the config

aleclarson commented on src/node/config.ts:
```diff
@@ -627,106 +624,75 @@ async function loadConfigFromBundledFile(
-      config.enableRollupPluginVue || plugin.enableRollupPluginVue
+function mergePlugin(config: ResolvedConfig, plugin: Plugin) {
+  if (typeof plugin === 'function') {
+    plugin = plugin(config)
```
I removed `changeConfig` in favor of plugin functions that receive a stable reference to the `config` object.

aleclarson commented on src/node/config.ts:
```diff
@@ -627,106 +624,75 @@ async function loadConfigFromBundledFile(
-  from: SFCTemplateCompileOptions['transformAssetUrls']
-): SFCTemplateCompileOptions['transformAssetUrls'] {
-  if (from === true) {
-    return to
```
@yyx990803 This line implies that an existing value of `vueTransformAssetUrls: false` should take precedence over a new value of `vueTransformAssetUrls: true`. Was this intended?

aleclarson committed:
fix: merge user config like a plugin

So that objects/arrays are shallow cloned, and to avoid the `vueTransformAssetUrls` special handling.

aleclarson committed:
fix: assign to `config.env` instead of cloning

aleclarson committed:
fix: resolve the `root` before merging plugins

aleclarson committed:
fix: let plugin functions return void

aleclarson committed:
refactor: move default values for `SharedConfig` into `resolveConfig` function

aleclarson committed:
refactor(build): use ??= syntax instead of cloning the config

aleclarson committed:
chore: remove klona
